### PR TITLE
Add `as.data.frame()` methods for `<epidist>` & `<multi_epidist>`

### PR DESCRIPTION
**Pull request by joshwlambert**
This PR addresses #249 by providing a method to dispatch the `as.data.frame()` generic for `<epidist>`, and by extension of looping over the `<epidist>` method, the `<multi_epidist>` classes. 

This is to provide {epiparameter} data in a tabular form to make it more easily applicable to pipelines, e.g. Tidyverse data analysis pipelines.

### CODE REVIEW

File: tests/testthat/test-epidist.R
Comment (pratikunterwegs) [inline tests/testthat/test-epidist.R:1151]: I would suggest defining the expected numbers of rows and columns as a variable and passing it here, as where these come from might be confusing for future maintainers - you could reuse the vector of expected column names below.

**Review by pratikunterwegs — APPROVED** [review]
This looks good to me @joshwlambert, nice and clean.

Just thinking through the use cases, it might be useful to provide the citation in a flat format, as the `<bibentry>` class (surprisingly) doesn't appear to support `[[`, making it more difficult to filter a data.frame of parameters by year or author, which might be relevant to comparative analyses.

We talked about taking on {tibble} as a dependency yesterday since this implementation creates data.frames with list columns, and I think that that is the right way to go - perhaps even in this PR. The function name might need changing in that case.

A tibble with list-columns will be more recognisable to users with Tidyverse training than a data.frame implementing the same. The printing of list-columns and large strings (such as `notes`) is better in a tibble too. Users would also probably use Tidyverse packages such as {purrr} to work with the list columns, which reduces the user-side cost of taking on {tibble} in {epiparameter}.

**Comment by joshwlambert** [comment]
> Just thinking through the use cases, it might be useful to provide the citation in a flat format, as the <bibentry> class (surprisingly) doesn't appear to support [[, making it more difficult to filter a data.frame of parameters by year or author, which might be relevant to comparative analyses.

By flat do you mean one `<data.frame>` column per `<bibentry>` element, or just unclass the `<bibentry>` when putting into the list column so it is just a list?

**Comment by pratikunterwegs** [comment]
> > Just thinking through the use cases, it might be useful to provide the citation in a flat format, as the  class (surprisingly) doesn't appear to support [[, making it more difficult to filter a data.frame of parameters by year or author, which might be relevant to comparative analyses.
> 
> By flat do you mean one `<data.frame>` column per `<bibentry>` element, or just unclass the `<bibentry>` when putting into the list column so it is just a list?

I meant the former, but the latter might also be viable if users are mostly using the citation data rather than the overall object and any related methods.

**Comment by joshwlambert** [comment]
> We talked about taking on {tibble} as a dependency yesterday since this implementation creates data.frames with list columns, and I think that that is the right way to go - perhaps even in this PR. The function name might need changing in that case.

> A tibble with list-columns will be more recognisable to users with Tidyverse training than a data.frame implementing the same. The printing of list-columns and large strings (such as notes) is better in a tibble too. Users would also probably use Tidyverse packages such as {purrr} to work with the list columns, which reduces the user-side cost of taking on {tibble} in {epiparameter}.

This is a nice idea, but I don't think it should be part of the `as.data.frame()` method, instead I propose making a future PR adding an `as_tibble()` method for `<epidist>` which would provide the formatting benefits you mentioned.

**Comment by avallecam** [comment]
> whether the list columns hard for users to understand and use?

My assessment is that the output of `<named list>` was not so intuitive to me on how to manage it to extract information to select specific entries. I needed to explore for a while how to unnest the data using {tidyr}. While doing this I discovered the existence of `tidyr::unnest_auto()` that unnested the data within, e.g., the `metadata` column into new columns for each observation (or row), making the `sample_size` column visible (whether unnesting as columns or row is the "auto" part, instead, I could also use `tidyr::unnest_wider()`). However, this method does not work for `citation` and `prob_distribution` with two different outputs.

Here is the code from my exploration:

``` r
# remotes::install_github("epiverse-trace/epiparameter@epidist2df")

library(epiparameter)
library(tidyverse)

# look for one distribution, but get four
epiparameter::epidist_db(
  disease = "covid",
  epi_dist = "serial"
) %>%
  epiparameter::list_distributions()
#> Returning 4 results that match the criteria (3 are parameterised). 
#> Use subset to filter by entry variables or single_epidist to return a single entry. 
#> To retrieve the short citation for each use the 'get_citation' function
#>    disease epi_distribution prob_distribution       author year
#> 1 COVID-19  serial interval              <NA> Muluneh .... 2021
#> 2 COVID-19  serial interval             lnorm Hiroshi .... 2020
#> 3 COVID-19  serial interval           weibull Hiroshi .... 2020
#> 4 COVID-19  serial interval              norm Lin Yang.... 2020

# I can use single_epidist = TRUE 
# but how much they differ in sample size? (or other variables)

# we can coerce multi_epidist to tibble
epiparameter::epidist_db(
  disease = "covid",
  epi_dist = "serial"
) %>%
  as_tibble()
#> Returning 4 results that match the criteria (3 are parameterised). 
#> Use subset to filter by entry variables or single_epidist to return a single entry. 
#> To retrieve the short citation for each use the 'get_citation' function
#> Warning: In as.data.frame.multi_epidist(value, stringsAsFactors = FALSE) :
#>  extra argument 'stringsAsFactors' will be disregarded
#> # A tibble: 4 × 10
#>   disease pathogen epi_distribution prob_distribution uncertainty  summary_stats
#>   <chr>   <chr>    <chr>            <I<list>>         <I<list>>    <I<list>>    
#> 1 COVID-… SARS-Co… serial interval  <lgl [1]>         <named list> <named list> 
#> 2 COVID-… SARS-Co… serial interval  <dist [1]>        <named list> <named list> 
#> 3 COVID-… SARS-Co… serial interval  <dist [1]>        <named list> <named list> 
#> 4 COVID-… SARS-Co… serial interval  <dist [1]>        <named list> <named list> 
#> # ℹ 4 more variables: citation <I<list>>, metadata <I<list>>,
#> #   method_assess <I<list>>, notes <chr>

# get a brief overview with glimpse()
epiparameter::epidist_db(
  disease = "covid",
  epi_dist = "serial"
) %>%
  as_tibble() %>% 
  glimpse()
#> Returning 4 results that match the criteria (3 are parameterised). 
#> Use subset to filter by entry variables or single_epidist to return a single entry. 
#> To retrieve the short citation for each use the 'get_citation' function
#> Warning: In as.data.frame.multi_epidist(value, stringsAsFactors = FALSE) :
#>  extra argument 'stringsAsFactors' will be disregarded
#> Rows: 4
#> Columns: 10
#> $ disease           <chr> "COVID-19", "COVID-19", "COVID-19", "COVID-19"
#> $ pathogen          <chr> "SARS-CoV-2", "SARS-CoV-2", "SARS-CoV-2", "SARS-CoV-…
#> $ epi_distribution  <chr> "serial interval", "serial interval", "serial interv…
#> $ prob_distribution <I<list>>           NA, lN(1.4, ...., Weibull(....,   N(4.6, 1…
#> $ uncertainty       <I<list>> NA, c(NA...., NA, c(NA...., NA, c(NA...., list(c…
#> $ summary_stats     <I<list>> 5.2, c(4...., 4.7, c(3...., 4.8, c(3...., c(`95`…
#> $ citation          <I<list>> list(lis...., list(lis...., list(lis...., list(l…
#> $ metadata          <I<list>> 3924, Mi...., 28, Mixe...., 18, Mixe...., 131, H…
#> $ method_assess     <I<list>> FALSE, F...., TRUE, TR...., TRUE, TR...., TRUE, …
#> $ notes             <chr> "This estimated mean serial interval is from a m…

# I can asume number within metadata 
# may be the sample size, but

# we can unnest list columns
# using unnest_auto() to have 
# a column named sample_size
epiparameter::epidist_db(
  disease = "covid",
  epi_dist = "serial"
) %>%
  as_tibble() %>% 
  unnest_auto(col = metadata) %>% 
  glimpse()
#> Returning 4 results that match the criteria (3 are parameterised). 
#> Use subset to filter by entry variables or single_epidist to return a single entry. 
#> To retrieve the short citation for each use the 'get_citation' function
#> Warning: In as.data.frame.multi_epidist(value, stringsAsFactors = FALSE) :
#>  extra argument 'stringsAsFactors' will be disregarded
#> Using `unnest_wider(metadata)`; elements have 6 names in common
#> Rows: 4
#> Columns: 15
#> $ disease           <chr> "COVID-19", "COVID-19", "COVID-19", "COVID-19"
#> $ pathogen          <chr> "SARS-CoV-2", "SARS-CoV-2", "SARS-CoV-2", "SARS-CoV-…
#> $ epi_distribution  <chr> "serial interval", "serial interval", "serial interv…
#> $ prob_distribution <I<list>>           NA, lN(1.4, ...., Weibull(....,   N(4.6, 1…
#> $ uncertainty       <I<list>> NA, c(NA...., NA, c(NA...., NA, c(NA...., list(c…
#> $ summary_stats     <I<list>> 5.2, c(4...., 4.7, c(3...., 4.8, c(3...., c(`95`…
#> $ citation          <I<list>> list(lis...., list(lis...., list(lis...., list(l…
#> $ sample_size       <int> 3924, 28, 18, 131
#> $ region            <chr> "Mixed", "Mixed", "Mixed", "Hubei province, China"
#> $ transmission_mode <chr> "natural_natural_human_to_human", "natural_natur…
#> $ vector            <chr> NA, NA, NA, NA
#> $ extrinsic         <lgl> FALSE, FALSE, FALSE, FALSE
#> $ inference_method  <chr> NA, "bayesian", "bayesian", "mle"
#> $ method_assess     <I<list>> FALSE, F...., TRUE, TR...., TRUE, TR...., TRUE, FA..…
#> $ notes             <chr> "This estimated mean serial interval is from a meta-…

# unnest_auto works for all list columns
# but not for
# - citation
# - prob_distribution

epiparameter::epidist_db(
  disease = "covid",
  epi_dist = "serial"
) %>%
  as_tibble() %>% 
  unnest_auto(col = citation) %>% 
  glimpse()
#> Returning 4 results that match the criteria (3 are parameterised). 
#> Use subset to filter by entry variables or single_epidist to return a single entry. 
#> To retrieve the short citation for each use the 'get_citation' function
#> Warning: In as.data.frame.multi_epidist(value, stringsAsFactors = FALSE) :
#>  extra argument 'stringsAsFactors' will be disregarded
#> Using `unnest_longer(citation, indices_include = FALSE)`; no element has names
#> Error in `vec_size()`:
#> ! `x` must be a vector, not a <bibentry> object.

epiparameter::epidist_db(
  disease = "covid",
  epi_dist = "serial"
) %>%
  as_tibble() %>% 
  unnest_auto(col = prob_distribution) %>% 
  glimpse()
#> Returning 4 results that match the criteria (3 are parameterised). 
#> Use subset to filter by entry variables or single_epidist to return a single entry. 
#> To retrieve the short citation for each use the 'get_citation' function
#> Warning: In as.data.frame.multi_epidist(value, stringsAsFactors = FALSE) :
#>  extra argument 'stringsAsFactors' will be disregarded
#> Using `unnest_longer(prob_distribution, indices_include = FALSE)`; no element has names
#> Rows: 4
#> Columns: 10
#> Error: C stack usage  15926144 is too close to the limit
```

<sup>Created on 2024-03-08 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by avallecam** [comment]
With respect to the data requirement that triggered my feature request, we work going on in the [parameter_tbl](https://github.com/epiverse-trace/epiparameter/tree/parameter_tbl) branch directly targets this issue nicely. I'm able to select any combination of authors and subset parameters, which can also help to try iterative calculations potentially using {purrr}

``` r
# remotes::install_github("epiverse-trace/epiparameter@parameter_tbl")

library(epiparameter)
library(tidyverse)

epidist_db(
  disease = "COVID-19",
  epi_dist = "incubation period"
) %>% 
  epiparameter::distribution_tbl()
#> Returning 15 results that match the criteria (11 are parameterised). 
#> Use subset to filter by entry variables or single_epidist to return a single entry. 
#> To retrieve the short citation for each use the 'get_citation' function
#> # Distribution table:
#> # A data frame:       15 × 7
#>    disease  pathogen epi_distribution prob_distribution author  year sample_size
#>    <chr>    <chr>    <chr>            <chr>             <chr>  <dbl>       <dbl>
#>  1 COVID-19 SARS-Co… incubation peri… <NA>              Men e…  2020          59
#>  2 COVID-19 SARS-Co… incubation peri… <NA>              Rai e…  2022        6241
#>  3 COVID-19 SARS-Co… incubation peri… <NA>              Alene…  2021        1453
#>  4 COVID-19 SARS-Co… incubation peri… weibull           Yang …  2020         178
#>  5 COVID-19 SARS-Co… incubation peri… <NA>              Elias…  2021       28675
#>  6 COVID-19 SARS-Co… incubation peri… weibull           Bui e…  2020          19
#>  7 COVID-19 SARS-Co… incubation peri… lnorm             McAlo…  2020        1357
#>  8 COVID-19 SARS-Co… incubation peri… lnorm             McAlo…  2020        1269
#>  9 COVID-19 SARS-Co… incubation peri… lnorm             Linto…  2020          52
#> 10 COVID-19 SARS-Co… incubation peri… lnorm             Linto…  2020         158
#> 11 COVID-19 SARS-Co… incubation peri… lnorm             Linto…  2020          52
#> 12 COVID-19 SARS-Co… incubation peri… lnorm             Lauer…  2020         181
#> 13 COVID-19 SARS-Co… incubation peri… lnorm             Lauer…  2020          99
#> 14 COVID-19 SARS-Co… incubation peri… lnorm             Lauer…  2020         108
#> 15 COVID-19 SARS-Co… incubation peri… lnorm             Lauer…  2020          73

epidist_db(
  disease = "COVID-19",
  epi_dist = "incubation period",
  author = "Linton",
  subset = sample_size == 158
)
#> Returning 1 results that match the criteria (1 are parameterised). 
#> Use subset to filter by entry variables or single_epidist to return a single entry. 
#> To retrieve the short citation for each use the 'get_citation' function
#> Disease: COVID-19
#> Pathogen: SARS-CoV-2
#> Epi Distribution: incubation period
#> Study: Linton N, Kobayashi T, Yang Y, Hayashi K, Akhmetzhanov A, Jung S, Yuan
#> B, Kinoshita R, Nishiura H (2020). "Incubation Period and Other
#> Epidemiological Characteristics of 2019 Novel Coronavirus Infections
#> with Right Truncation: A Statistical Analysis of Publicly Available
#> Case Data." _Journal of Clinical Medicine_. doi:10.3390/jcm9020538
#> <https://doi.org/10.3390/jcm9020538>.
#> Distribution: lnorm
#> Parameters:
#>   meanlog: 1.611
#>   sdlog: 0.472
```

<sup>Created on 2024-03-08 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Review by avallecam — CHANGES_REQUESTED** [review]
using `unnest_wider()` for list columns with `<named list>` objects within is introduced in the [unnesting episode from r4ds](https://r4ds.hadley.nz/rectangling.html#unnesting). Even though, this interface was not familiar to me, we can expect this to be familiar to coming users and refer to this material as a prerequisite. Here I copy a reprex from the episode:

``` r
library(tidyverse)

df1 <- tribble(
  ~x, ~y,
  1, list(a = 11, b = 12),
  2, list(a = 21, b = 22),
  3, list(a = 31, b = 32),
)

df1
#> # A tibble: 3 × 2
#>       x y               
#>   <dbl> <list>          
#> 1     1 <named list [2]>
#> 2     2 <named list [2]>
#> 3     3 <named list [2]>

df1 |> 
  unnest_wider(y)
#> # A tibble: 3 × 3
#>       x     a     b
#>   <dbl> <dbl> <dbl>
#> 1     1    11    12
#> 2     2    21    22
#> 3     3    31    32
```

<sup>Created on 2024-03-08 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

I would only request to suggest how to manage the output for `citation`, or if we should, especially for the <person> object. e.g., how to get author names as characters? (happy to re-assess if this does not require code edits)

``` r
# remotes::install_github("epiverse-trace/epiparameter@epidist2df")

library(epiparameter)
library(tidyverse)

epidist_input <- epiparameter::epidist_db(
  disease = "covid",
  epi_dist = "serial"
)
#> Returning 4 results that match the criteria (3 are parameterised). 
#> Use subset to filter by entry variables or single_epidist to return a single entry. 
#> To retrieve the short citation for each use the 'get_citation' function

epidist_tibble <- epidist_input %>%
  as_tibble()
#> Warning: In as.data.frame.multi_epidist(value, stringsAsFactors = FALSE) :
#>  extra argument 'stringsAsFactors' will be disregarded

epidist_tibble
#> # A tibble: 4 × 10
#>   disease pathogen epi_distribution prob_distribution uncertainty  summary_stats
#>   <chr>   <chr>    <chr>            <I<list>>         <I<list>>    <I<list>>    
#> 1 COVID-… SARS-Co… serial interval  <lgl [1]>         <named list> <named list> 
#> 2 COVID-… SARS-Co… serial interval  <dist [1]>        <named list> <named list> 
#> 3 COVID-… SARS-Co… serial interval  <dist [1]>        <named list> <named list> 
#> 4 COVID-… SARS-Co… serial interval  <dist [1]>        <named list> <named list> 
#> # ℹ 4 more variables: citation <I<list>>, metadata <I<list>>,
#> #   method_assess <I<list>>, notes <chr>

# works
epidist_tibble %>% 
  select(disease,epi_distribution,citation,metadata) %>% 
  unnest_wider(col = metadata)
#> # A tibble: 4 × 9
#>   disease  epi_distribution citation     sample_size region    transmission_mode
#>   <chr>    <chr>            <I<list>>          <int> <chr>     <chr>            
#> 1 COVID-19 serial interval  <named list>        3924 Mixed     natural_natural_…
#> 2 COVID-19 serial interval  <named list>          28 Mixed     natural_natural_…
#> 3 COVID-19 serial interval  <named list>          18 Mixed     natural_natural_…
#> 4 COVID-19 serial interval  <named list>         131 Hubei pr… natural_natural_…
#> # ℹ 3 more variables: vector <chr>, extrinsic <lgl>, inference_method <chr>

# this required unnest() only, not unnest_wider() -------------------------------------------------
epidist_tibble %>% 
  unnest(prob_distribution)
#> # A tibble: 4 × 10
#>   disease pathogen epi_distribution prob_distribution uncertainty  summary_stats
#>   <chr>   <chr>    <chr>                       <dist> <I<list>>    <I<list>>    
#> 1 COVID-… SARS-Co… serial interval                 NA <named list> <named list> 
#> 2 COVID-… SARS-Co… serial interval      lN(1.4, 0.32) <named list> <named list> 
#> 3 COVID-… SARS-Co… serial interval  Weibull(2.2, 5.4) <named list> <named list> 
#> 4 COVID-… SARS-Co… serial interval         N(4.6, 19) <named list> <named list> 
#> # ℹ 4 more variables: citation <I<list>>, metadata <I<list>>,
#> #   method_assess <I<list>>, notes <chr>

# Should we deal with the citation column this way? -----------------------------------------------
epidist_tibble %>%
  select(disease,epi_distribution,citation) %>% 
  unnest(citation) %>% 
  unnest_longer(citation) %>% 
  filter(citation_id != "author") %>% 
  unnest(citation)
#> # A tibble: 20 × 4
#>    disease  epi_distribution citation                                citation_id
#>    <chr>    <chr>            <chr>                                   <chr>      
#>  1 COVID-19 serial interval  2021                                    year       
#>  2 COVID-19 serial interval  Serial interval and incubation period … title      
#>  3 COVID-19 serial interval  BMC Infectious Diseases                 journal    
#>  4 COVID-19 serial interval  10.1186/s12879-021-05950-x              doi        
#>  5 COVID-19 serial interval  33706702                                pmid       
#>  6 COVID-19 serial interval  2020                                    year       
#>  7 COVID-19 serial interval  Serial interval of novel coronavirus (… title      
#>  8 COVID-19 serial interval  International Journal of Infectious Di… journal    
#>  9 COVID-19 serial interval  10.1016/j.ijid.2020.02.060              doi        
#> 10 COVID-19 serial interval  32145466                                pmid       
#> 11 COVID-19 serial interval  2020                                    year       
#> 12 COVID-19 serial interval  Serial interval of novel coronavirus (… title      
#> 13 COVID-19 serial interval  International Journal of Infectious Di… journal    
#> 14 COVID-19 serial interval  10.1016/j.ijid.2020.02.060              doi        
#> 15 COVID-19 serial interval  32145466                                pmid       
#> 16 COVID-19 serial interval  2020                                    year       
#> 17 COVID-19 serial interval  Estimation of incubation period and se… title      
#> 18 COVID-19 serial interval  Epidemiology and Infection              journal    
#> 19 COVID-19 serial interval  10.1017/S0950268820001338               doi        
#> 20 COVID-19 serial interval  32594928                                pmid

epidist_tibble %>%
  select(disease,epi_distribution,citation) %>% 
  unnest(citation) %>% 
  unnest_longer(citation) %>% 
  filter(citation_id == "author")
#> # A tibble: 4 × 4
#>   disease  epi_distribution citation     citation_id
#>   <chr>    <chr>            <named list> <chr>      
#> 1 COVID-19 serial interval  <person>     author     
#> 2 COVID-19 serial interval  <person>     author     
#> 3 COVID-19 serial interval  <person>     author     
#> 4 COVID-19 serial interval  <person>     author
```

<sup>Created on 2024-03-08 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

File: R/epidist.R
Comment (avallecam) [inline R/epidist.R:1071]: is there a reason why you are using the `I(list(...))` and getting `<I<list>>` columns? this feature is new to me.

Comment (joshwlambert) [reply R/epidist.R:1071]: Without using the `I()` function the `data.frame()` function will unnest the list into several columns. The `I()` converts to `<AsIs>` as keeps the objects together and stored as a list column.

File: tests/testthat/test-epidist.R
Comment (joshwlambert) [reply tests/testthat/test-epidist.R:1151]: I've just implemented this idea using variables to hold the number of columns and rows expected, however it feels a bit over the top. I think that the expectation is still readable as a comparison of the dimensions to the vector. Will leave as is for now.

**Comment by joshwlambert** [comment]
Thanks for the comments. I will now merge this PR. Some of the latter comments are more a response to the developments on the `parameter_tbl` branch, so I will revisit them once a PR is open for merged that branch into `main`.

**Pull request merged**
